man/bppdist.1
.\" SPDX-FileCopyrightText: The Bio++ Development Group
.\"
.\" SPDX-License-Identifier: CECILL-2.1

.TH BPPDIST 1 LOCAL

.SH NAME

bppdist - Distance methods for phylogenetic reconstruction with Bio++

.SH SYNOPSIS

.B bppdist [options]

.SH AVAILABILITY

All UNIX flavors

.SH DESCRIPTION

bppdist performs phylogenetic reconstruction using distance methods (like neighbor joining or bioNJ).

.SH OPTIONS

You should refer to 'info bppsuite' or to the online manual of bppsuite for a complete list of available options.

.TP 5

--noninteractive

generates output for redirection in a file.

.TP

param=file

reads a file for loading options

.SH AUTHOR

The Bio++ Development Team.


ChangeLog
19/03/18 -*- Version 2.4.0 -*-

19/02/18 Julien Dutheil
* Bug fixed in bppPopStats: dN and dS were inverted!

10/12/17 -*- Version 2.3.2 -*-

06/11/17 Julien Dutheil
* Added estimation of kappa to bppPopStats + bugs fixed

06/06/17 -*- Version 2.3.1 -*- 

10/05/17 -*- Version 2.3.0 -*- 

21/12/16 Julien Dutheil
* BppPhySamp migrated to its own package.

17/12/16 Julien Dutheil
* Improved bppPopStats, outgroup analyses, MK table and output to file.

06/10/16 Julien Dutheil
* bppSeqGen now takes Hudson's MS ARG files as input.
* new input.tree.scale option to adjust branch lengths.

22/05/16 Julien Dutheil
* bpppopstats now computes codon statistics per site (e.g. PiN, PiS)
* Added example for bpppopstats.

29/06/15 Julien Dutheil
* New program bppPopStats

28/01/15 Julien Dutheil
* PhySamp now outputs the correct tree!

25/11/14 Julien Dutheil
* bppalnscore output contiguous segments in mase selection.
* bppalnscore site selection works with codon alphabets.

24/11/14 Julien Dutheil
* bppSeqMan can now read alignments and perform site selection priori to any analysis.

28/09/14 -*- Version 2.2.0 -*- 

16/09/14 Julien Dutheil
* bppSeqGen supports generic characters as ancestral states in info file (bug #87).

12/08/14 Julien Dutheil
* bppSeqGen supports simulation under Markov-Modulated models.

06/06/14 Julien Dutheil
* bppPhySamp now output trees as well.

17/04/14 Julien Dutheil
* bppSeqMan: Translate uses global genetic_code argument instead of local 'code' argument.
* All programs use the --seed argument to set random seed for reproduceable results.

08/03/13 -*- Version 0.8.0 -*-

22/01/13 Julien Dutheil
* Compiles with -Wconversion

14/01/13 Mathieu Groussin
* Added COaLA model for proteins in bppML and bppSeqGen.

23/11/12 Laurent Guéguen
* New bppMixedLikelihood program

08/02/12 -*- Version 0.7.0 -*-

21/12/11 Julien Dutheil
* New bppAlnScore program
* Improved bppSeqMan: protein translation now works + more stop codons removal options.

17/06/11 Julien Dutheil
* Improved bppancestor, also allows to compute frequencies for leaves.

11/06/11 -*- Version 0.6.2 -*-
* Small bug fixed + new packages.

03/06/10 -*- Version 0.6.0 -*-
* Compatibility with Bio++ 2.0.0.

03/06/10 Julien Dutheil
* New method in bppPhySamp + new, faster, implementation.

25/03/10 -*- Version 0.5.0 -*-

22/08/09 Julien Dutheil
* Added CMake configuration files, with twick for building the manual (assumes
		texinfo is installed).

22/07/09 Julien Dutheil
* Added program bppTreeDraw.

30/06/09 Julien Dutheil
* BppAncestor: now output expected ancestral states frequencies.
* BppAncestor: new method to compute EB estimates of ancestral states frequencies.

10/06/09 -*- Version 0.4.0 -*-

22/05/09 Julien Dutheil
* Keyval syntax for sequence format description.
* Documentation and example update.
* SeqMan now uses the SequenceApplicationTools method.

15/05/09 Julien Dutheil
* Compatibility update with the libs, new keyval syntax for model
specification.
* Documentation up to date.

02/02/09 Julien Dutheil
* bppConsense now outputs the number of leaves of the input tree.
* New progress bar in bppConsense.

08/01/09 Julien Dutheil
* Added texinfo documentation.
* More bootstrap options in BppDist (the same as in BppML).

25/09/08 -*- Version 0.3.1 -*-

07/12/08 Bastien Boussau & Julien Dutheil
* Bug removed in non-homogeneous models with covarions.

09/10/08 Julien Dutheil
* New action 'invert' in bppSeqMan.

25/09/08 -*- Version 0.3.0 -*-

20/09/08 Julien Dutheil
* New program bppAncestor to reconstruct acestral states and rates.
* New +F models from PhylLib. 

02/07/08 Laurent Guegen
* Bug fixed in bppML: frequencies can be computed from the data with
non-homogeneous models.

19/06/08 Julien Dutheil
* Shell variables are now 'global' in option files.

24/04/08 -*- Version 0.2.3 -*-
* This version requires the CVS version of Bio++ to be built.

03/03/08 Julien Dutheil
* Bug fixed in keep_complete method of BppSeqMan.

27/03/08 Julien Dutheil
* Bug fixed when printing parameters of non-homogeneous models in bppML.

22/01/08 -*- Version 0.2.2 -*-

28/02/08 Celine Scronavacca
* Bug fixed in ReRoot.

22/01/08 -*- Version 0.2.1 -*-

06/02/08 Celine Scornavacca
* Bug fixed in ReRoot + correction of version number in bppSeqGen.

22/01/08 -*- Version 0.2.0 -*-

13/01/08 Julien Dutheil
* More options in bppml: possibility to chose between simple and double
recursion and data compression method.
* Simple recursion method improved.

11/01/08 Julien Dutheil
* Compatibility update
* bppML now output more handy parameter values.

31/12/07 Julien Dutheil
* New program Phylo Sampler.

07/11/07 Julien Dutheil
* Compatibility update with Bio++ cvs.
* Bug fixed in BppSeqGen.
* BppSeqGen can now simulate data from a set of trees (ancestral
  recombination graph).

25/10/07 Julien Dutheil
* Several bug fixed in BppML and BppSeqGen with non-homogeneous models.
* Examples updated + new example option files for non-homogeneous models.

05/10/07 Bastien Boussau & Julien Dutheil
* Non-homogenous sequence simulation in BppSeqGen.
* Model options in files have changed to suit the non-homogeneous case.
  The covarion option now is fully part of the model options, following the
	new syntax 'model=modelname+covarionname', eg 'T92+G2001'.

03/10/07 Julien Dutheil
* New program Bio++ Sequence Manipulator

21/09/07 Julien Dutheil
* Bug fixed in BppDist: branch length are indeed not estimated when parameters
  to ignore contains "BrLen".
* Stop condition when using iterations is now based on topological distance
  instead of likelihood comparisons (phyllib).
* New option optimization.topology_nni.method = [fast|better|phyml]

17/06/07 -*- Version 0.1.0 -*-

16/07/06 Julien Dutheil
* New option bootstrap.verbose for bppML.
* Bug fixed in bootstrap computation.

06/06/06 Julien Dutheil
* New program bppConsense.

05/05/07 Julien Dutheil
* First alpha release of bppsuite.

-*- Old bppML ChangeLog -*-

23/04/07 Julien Dutheil

bppSuite/bppDist.cpp
// SPDX-FileCopyrightText: The Bio++ Development Group
//
// SPDX-License-Identifier: CECILL-2.1

// From the STL:
#include <iostream>
#include <iomanip>

using namespace std;

// From bpp-core:
#include <Bpp/Version.h>
#include <Bpp/Numeric/Prob/DiscreteDistribution.h>
#include <Bpp/Io/FileTools.h>
#include <Bpp/Text/TextTools.h>

// From bpp-seq:
#include <Bpp/Seq/Alphabet/Alphabet.h>
#include <Bpp/Seq/Container/VectorSiteContainer.h>
#include <Bpp/Seq/Container/SiteContainerTools.h>
#include <Bpp/Seq/App/BppSequenceApplication.h>
#include <Bpp/Seq/SiteTools.h>
#include <Bpp/Seq/App/SequenceApplicationTools.h>
#include <Bpp/Text/KeyvalTools.h>


// From PhylLib:
#include <Bpp/Phyl/Tree/Tree.h>
#include <Bpp/Phyl/PatternTools.h>
#include <Bpp/Phyl/App/PhylogeneticsApplicationTools.h>
#include <Bpp/Phyl/Io/Newick.h>
#include <Bpp/Phyl/Io/IoDistanceMatrixFactory.h>
#include <Bpp/Phyl/Distance/DistanceEstimation.h>
#include <Bpp/Phyl/Distance/PGMA.h>
#include <Bpp/Phyl/Distance/BioNJ.h>
#include <Bpp/Phyl/OptimizationTools.h>
#include <Bpp/Phyl/Model/RateDistribution/ConstantRateDistribution.h>

using namespace bpp;

void help()
{
  (*ApplicationTools::message << "__________________________________________________________________________").endLine();
  (*ApplicationTools::message << "bppdist parameter1_name=parameter1_value parameter2_name=parameter2_value").endLine();
  (*ApplicationTools::message << "      ... param=option_file").endLine();
  (*ApplicationTools::message).endLine();
  (*ApplicationTools::message << "  Refer to the Bio++ Program Suite Manual for a list of available options.").endLine();
  (*ApplicationTools::message << "__________________________________________________________________________").endLine();
}

int main(int args, char** argv)
{
  cout << "******************************************************************" << endl;
  cout << "*              Bio++ Distance Methods, version " << BPP_VERSION << "             *" << endl;
  cout << "* Author: J. Dutheil                        Created     05/05/07 *" << endl;
  cout << "*                                           Last Modif. " << BPP_REL_DATE << " *" << endl;
  cout << "******************************************************************" << endl;
  cout << endl;

  if (args == 1)
  {
    help();
    return 0;
  }

  try
  {
    BppSequenceApplication bppdist(args, argv, "BppDist");
    bppdist.startTimer();

    std::shared_ptr<const Alphabet> alphabet = bppdist.getAlphabet();

    /// GeneticCode

    auto gCode(bppdist.getGeneticCode(alphabet));

    // sites

    auto allSites = SequenceApplicationTools::getSiteContainer(alphabet, bppdist.getParams());

    shared_ptr<VectorSiteContainer> sites = SequenceApplicationTools::getSitesToAnalyse(*allSites, bppdist.getParams());

    // model

    map<string, string> unparsedparams;

    std::shared_ptr<BranchModelInterface> model = PhylogeneticsApplicationTools::getBranchModel(alphabet, gCode, sites, bppdist.getParams(), unparsedparams);

    if (!model)
      ApplicationTools::displayMessage("bppDist available only for regular transition models, not the given one.");


    std::shared_ptr<DiscreteDistributionInterface> rDist;
    if (model->getNumberOfStates() > model->getAlphabet()->getSize())
    {
      // Markov-modulated Markov model!
      rDist = std::shared_ptr<ConstantRateDistribution>();
    }
    else
    {
      rDist = std::shared_ptr<DiscreteDistributionInterface>(PhylogeneticsApplicationTools::getRateDistribution(bppdist.getParams()));
    }

    DistanceEstimation distEstimation(model, rDist, sites, 1, false);

    string method = ApplicationTools::getStringParameter("method", bppdist.getParams(), "nj");
    ApplicationTools::displayResult("Tree reconstruction method", method);

    unique_ptr<TreeTemplate<Node>> tree;
    unique_ptr<AgglomerativeDistanceMethodInterface> distMethod;
    if (method == "wpgma")
    {
      distMethod = make_unique<PGMA>(true);
    }
    else if (method == "upgma")
    {
      distMethod = make_unique<PGMA>(false);
    }
    else if (method == "nj")
    {
      auto nj = make_unique<NeighborJoining>();
      nj->outputPositiveLengths(true);
      distMethod = std::move(nj);
    }
    else if (method == "bionj")
    {
      auto bionj = make_unique<BioNJ>();
      bionj->outputPositiveLengths(true);
      distMethod = std::move(bionj);
    }
    else
      throw Exception("Unknown tree reconstruction method.");

    string type = ApplicationTools::getStringParameter("optimization.method", bppdist.getParams(), "init");
    ApplicationTools::displayResult("Model parameters estimation method", type);
    if (type == "init")
      type = OptimizationTools::DISTANCEMETHOD_INIT;
    else if (type == "pairwise")
      type = OptimizationTools::DISTANCEMETHOD_PAIRWISE;
    else if (type == "iterations")
      type = OptimizationTools::DISTANCEMETHOD_ITERATIONS;
    else
      throw Exception("Unknown parameter estimation procedure '" + type + "'.");

    unsigned int optVerbose = ApplicationTools::getParameter<unsigned int>("optimization.verbose", bppdist.getParams(), 2);

    string mhPath = ApplicationTools::getAFilePath("optimization.message_handler", bppdist.getParams(), false, false);
    auto messenger = shared_ptr<OutputStream>(
          (mhPath == "none") ? 0 :
          (mhPath == "std") ? ApplicationTools::message.get() :
          new StlOutputStream(make_unique<ofstream>(mhPath.c_str(), ios::out)));

    ApplicationTools::displayResult("Message handler", mhPath);

    string prPath = ApplicationTools::getAFilePath("optimization.profiler", bppdist.getParams(), false, false);
    auto profiler = shared_ptr<OutputStream>(
          (prPath == "none") ? 0 :
          (prPath == "std") ? ApplicationTools::message.get() :
          new StlOutputStream(make_unique<ofstream>(prPath.c_str(), ios::out)));

    if (profiler)
      profiler->setPrecision(20);
    ApplicationTools::displayResult("Profiler", prPath);

    // Should I ignore some parameters?
    ParameterList allParameters = model->getParameters();
    allParameters.addParameters(rDist->getParameters());
    ParameterList parametersToIgnore;
    string paramListDesc = ApplicationTools::getStringParameter("optimization.ignore_parameter", bppdist.getParams(), "", "", true, false);
    bool ignoreBrLen = false;
    StringTokenizer st(paramListDesc, ",");
    while (st.hasMoreToken())
    {
      try
      {
        string param = st.nextToken();
        if (param == "BrLen")
          ignoreBrLen = true;
        else
        {
          if (allParameters.hasParameter(param))
          {
            Parameter* p = &allParameters.parameter(param);
            parametersToIgnore.addParameter(*p);
          }
          else
            ApplicationTools::displayWarning("Parameter '" + param + "' not found.");
        }
      }
      catch (ParameterNotFoundException& pnfe)
      {
        ApplicationTools::displayError("Parameter '" + pnfe.parameter() + "' not found, and so can't be ignored!");
      }
    }

    unsigned int nbEvalMax = ApplicationTools::getParameter<unsigned int>("optimization.max_number_f_eval", bppdist.getParams(), 1000000);
    ApplicationTools::displayResult("Max # ML evaluations", TextTools::toString(nbEvalMax));

    double tolerance = ApplicationTools::getDoubleParameter("optimization.tolerance", bppdist.getParams(), .000001);
    ApplicationTools::displayResult("Tolerance", TextTools::toString(tolerance));

    // Here it is:
    ofstream warn("warnings", ios::out);
    ApplicationTools::warning = std::shared_ptr<OutputStream>(dynamic_cast<OutputStream*>(new StlOutputStreamWrapper(&warn)));
    tree = OptimizationTools::buildDistanceTree(distEstimation, *distMethod, parametersToIgnore, !ignoreBrLen, type, tolerance, nbEvalMax, profiler, messenger, optVerbose);
    warn.close();
//    delete ApplicationTools::warning;
    ApplicationTools::warning = ApplicationTools::message;

    string matrixPath = ApplicationTools::getAFilePath("output.matrix.file", bppdist.getParams(), false, false, "", false);
    if (matrixPath != "none")
    {
      ApplicationTools::displayResult("Output matrix file", matrixPath);
      string matrixFormat = ApplicationTools::getAFilePath("output.matrix.format", bppdist.getParams(), false, false, "", false);
      string format = "";
      bool extended = false;
      std::map<std::string, std::string> unparsedArguments_;
      KeyvalTools::parseProcedure(matrixFormat, format, unparsedArguments_);
      if (unparsedArguments_.find("type") != unparsedArguments_.end())
      {
        if (unparsedArguments_["type"] == "extended")
        {
          extended = true;
        }
        else if (unparsedArguments_["type"] == "classic")
          extended = false;
        else
          ApplicationTools::displayWarning("Argument '" +
              unparsedArguments_["type"] + "' for parameter 'Phylip#type' is unknown. " +
              "Default used instead: not extended.");
      }
      else
        ApplicationTools::displayWarning("Argument 'Phylip#type' not found. Default used instead: not extended.");

      ODistanceMatrix* odm = IODistanceMatrixFactory().createWriter(IODistanceMatrixFactory::PHYLIP_FORMAT, extended);
      odm->writeDistanceMatrix(*distEstimation.getMatrix(), matrixPath, true);
      delete odm;
    }
    PhylogeneticsApplicationTools::writeTree(*tree, bppdist.getParams());

    // Output some parameters:
    if (type == OptimizationTools::DISTANCEMETHOD_ITERATIONS)
    {
      // Write parameters to screen:
      ParameterList parameters = model->getParameters();
      for (unsigned int i = 0; i < parameters.size(); i++)
      {
        ApplicationTools::displayResult(parameters[i].getName(), TextTools::toString(parameters[i].getValue()));
      }
      parameters = rDist->getParameters();
      for (unsigned int i = 0; i < parameters.size(); i++)
      {
        ApplicationTools::displayResult(parameters[i].getName(), TextTools::toString(parameters[i].getValue()));
      }
      // Write parameters to file:
      string parametersFile = ApplicationTools::getAFilePath("output.estimates", bppdist.getParams(), false, false);
      if (parametersFile != "none")
      {
        ofstream out(parametersFile.c_str(), ios::out);
        parameters = model->getParameters();
        for (unsigned int i = 0; i < parameters.size(); i++)
        {
          out << parameters[i].getName() << " = " << parameters[i].getValue() << endl;
        }
        parameters = rDist->getParameters();
        for (unsigned int i = 0; i < parameters.size(); i++)
        {
          out << parameters[i].getName() << " = " << parameters[i].getValue() << endl;
        }
        out.close();
      }
    }

    // Bootstrap:
    unsigned int nbBS = ApplicationTools::getParameter<unsigned int>("bootstrap.number", bppdist.getParams(), 0);
    if (nbBS > 0)
    {
      ApplicationTools::displayResult("Number of bootstrap samples", TextTools::toString(nbBS));
      bool approx = ApplicationTools::getBooleanParameter("bootstrap.approximate", bppdist.getParams(), true);
      ApplicationTools::displayResult("Use approximate bootstrap", TextTools::toString(approx ? "yes" : "no"));
      if (approx)
      {
        type = OptimizationTools::DISTANCEMETHOD_INIT;
        parametersToIgnore = allParameters;
        ignoreBrLen = true;
      }
      bool bootstrapVerbose = ApplicationTools::getBooleanParameter("bootstrap.verbose", bppdist.getParams(), false, "", true, false);

      string bsTreesPath = ApplicationTools::getAFilePath("bootstrap.output.file", bppdist.getParams(), false, false);
      ofstream* out = NULL;
      if (bsTreesPath != "none")
      {
        ApplicationTools::displayResult("Bootstrap trees stored in file", bsTreesPath);
        out = new ofstream(bsTreesPath.c_str(), ios::out);
      }
      Newick newick;

      vector<std::unique_ptr<Tree>> bsTrees(nbBS);
      ApplicationTools::displayTask("Bootstrapping", true);
      for (unsigned int i = 0; i < nbBS; i++)
      {
        ApplicationTools::displayGauge(i, nbBS - 1, '=');
        shared_ptr<VectorSiteContainer> sample = SiteContainerTools::bootstrapSites(*sites);
        auto tm = dynamic_pointer_cast<TransitionModelInterface>(model);
        if (approx && tm)
          tm->setFreqFromData(*sample);
        distEstimation.setData(sample);
        bsTrees[i] = OptimizationTools::buildDistanceTree(
              distEstimation,
              *distMethod,
              parametersToIgnore,
              ignoreBrLen,
              type,
              tolerance,
              nbEvalMax,
              NULL,
              NULL,
              (bootstrapVerbose ? 1 : 0)
              );
        if (out && i == 0)
          newick.writeTree(*bsTrees[i], bsTreesPath, true);
        if (out && i >  0)
          newick.writeTree(*bsTrees[i], bsTreesPath, false);
      }
      if (out)
        out->close();
      if (out)
        delete out;
      ApplicationTools::displayTaskDone();
      ApplicationTools::displayTask("Compute bootstrap values");
      TreeTools::computeBootstrapValues(*tree, bsTrees);
      ApplicationTools::displayTaskDone();

      // Write resulting tree:
      PhylogeneticsApplicationTools::writeTree(*tree, bppdist.getParams());
    }

    bppdist.done();
  }

  catch (exception& e)
  {
    cout << e.what() << endl;
    return 1;
  }

  return 0;
}
